### ISSUE TITLE
Wrong prometheus metrics reports while using horizontally placed nodes in an ensemble model DAG

### ISSUE THREAD

**Issue by julienripoche** [triton-inference-server/server#5296]
Hi triton server team !
Thanks for doing a great job ;)

Here is small prometheus report error that came to me while using ensemble model.

**Description**
I've implemented an object detection model as an ensemble model on tritonserver like this : 
```mermaid
  graph TD;
      backbone-->anchor_generator;
      backbone-->head_1;
      anchor_generator-->postprocess_detections;
      head_1-->postprocess_detections;
```
where all models are tensorRT engines but `postprocess_detections` which is an ONNX model.

The ensemble model that interests me has actually 9 heads like this :

```mermaid
  graph TD;
      backbone-->anchor_generator;
      backbone-->head_1;
      backbone-->head_2;
      backbone-->head_3;
      backbone-->head_4;
      backbone-->head_5;
      backbone-->head_6;
      backbone-->head_7;
      backbone-->head_8;
      backbone-->head_9;
      anchor_generator-->postprocess_detections_1;
      anchor_generator-->postprocess_detections_2;
      anchor_generator-->postprocess_detections_3;
      anchor_generator-->postprocess_detections_4;
      anchor_generator-->postprocess_detections_5;
      anchor_generator-->postprocess_detections_6;
      anchor_generator-->postprocess_detections_7;
      anchor_generator-->postprocess_detections_8;
      anchor_generator-->postprocess_detections_9;
      head_1-->postprocess_detections_1;
      head_2-->postprocess_detections_2;
      head_3-->postprocess_detections_3;
      head_4-->postprocess_detections_4;
      head_5-->postprocess_detections_5;
      head_6-->postprocess_detections_6;
      head_7-->postprocess_detections_7;
      head_8-->postprocess_detections_8;
      head_9-->postprocess_detections_9;
```
where postprocess_detections is the same model but run 9 times with different inputs.

In the first ensemble model with 1 head, the prometheus metrics are like this :

```
# HELP nv_inference_request_success Number of successful inference requests, all batch sizes
# TYPE nv_inference_request_success counter
nv_inference_request_success{model="model_head_1",version="1"} 100.000000
nv_inference_request_success{model="head_3",version="1"} 0.000000
nv_inference_request_success{model="anchor_generator",version="1"} 100.000000
nv_inference_request_success{model="head_4",version="1"} 0.000000
nv_inference_request_success{model="postprocess_detections",version="1"} 100.000000
nv_inference_request_success{model="head_1",version="1"} 100.000000
nv_inference_request_success{model="head_5",version="1"} 0.000000
nv_inference_request_success{model="head_6",version="1"} 0.000000
nv_inference_request_success{model="head_2",version="1"} 0.000000
nv_inference_request_success{model="head_7",version="1"} 0.000000
nv_inference_request_success{model="backbone",version="1"} 100.000000
nv_inference_request_success{model="head_8",version="1"} 0.000000
nv_inference_request_success{model="head_9",version="1"} 0.000000
nv_inference_request_success{model="model_all_heads",version="1"} 0.000000
# HELP nv_inference_request_failure Number of failed inference requests, all batch sizes
# TYPE nv_inference_request_failure counter
nv_inference_request_failure{model="model_head_1",version="1"} 0.000000
nv_inference_request_failure{model="head_3",version="1"} 0.000000
nv_inference_request_failure{model="anchor_generator",version="1"} 0.000000
nv_inference_request_failure{model="head_4",version="1"} 0.000000
nv_inference_request_failure{model="postprocess_detections",version="1"} 0.000000
nv_inference_request_failure{model="head_1",version="1"} 0.000000
nv_inference_request_failure{model="head_5",version="1"} 0.000000
nv_inference_request_failure{model="head_6",version="1"} 0.000000
nv_inference_request_failure{model="head_2",version="1"} 0.000000
nv_inference_request_failure{model="head_7",version="1"} 0.000000
nv_inference_request_failure{model="backbone",version="1"} 0.000000
nv_inference_request_failure{model="head_8",version="1"} 0.000000
nv_inference_request_failure{model="head_9",version="1"} 0.000000
nv_inference_request_failure{model="model_all_heads",version="1"} 0.000000
# HELP nv_inference_count Number of inferences performed (does not include cached requests)
# TYPE nv_inference_count counter
nv_inference_count{model="model_head_1",version="1"} 100.000000
nv_inference_count{model="head_3",version="1"} 0.000000
nv_inference_count{model="anchor_generator",version="1"} 100.000000
nv_inference_count{model="head_4",version="1"} 0.000000
nv_inference_count{model="postprocess_detections",version="1"} 100.000000
nv_inference_count{model="head_1",version="1"} 100.000000
nv_inference_count{model="head_5",version="1"} 0.000000
nv_inference_count{model="head_6",version="1"} 0.000000
nv_inference_count{model="head_2",version="1"} 0.000000
nv_inference_count{model="head_7",version="1"} 0.000000
nv_inference_count{model="backbone",version="1"} 100.000000
nv_inference_count{model="head_8",version="1"} 0.000000
nv_inference_count{model="head_9",version="1"} 0.000000
nv_inference_count{model="model_all_heads",version="1"} 0.000000
# HELP nv_inference_exec_count Number of model executions performed (does not include cached requests)
# TYPE nv_inference_exec_count counter
nv_inference_exec_count{model="model_head_1",version="1"} 100.000000
nv_inference_exec_count{model="head_3",version="1"} 0.000000
nv_inference_exec_count{model="anchor_generator",version="1"} 100.000000
nv_inference_exec_count{model="head_4",version="1"} 0.000000
nv_inference_exec_count{model="postprocess_detections",version="1"} 100.000000
nv_inference_exec_count{model="head_1",version="1"} 100.000000
nv_inference_exec_count{model="head_5",version="1"} 0.000000
nv_inference_exec_count{model="head_6",version="1"} 0.000000
nv_inference_exec_count{model="head_2",version="1"} 0.000000
nv_inference_exec_count{model="head_7",version="1"} 0.000000
nv_inference_exec_count{model="backbone",version="1"} 100.000000
nv_inference_exec_count{model="head_8",version="1"} 0.000000
nv_inference_exec_count{model="head_9",version="1"} 0.000000
nv_inference_exec_count{model="model_all_heads",version="1"} 0.000000
# HELP nv_inference_request_duration_us Cumulative inference request duration in microseconds (includes cached requests)
# TYPE nv_inference_request_duration_us counter
nv_inference_request_duration_us{model="model_head_1",version="1"} 5557899.000000
nv_inference_request_duration_us{model="head_3",version="1"} 0.000000
nv_inference_request_duration_us{model="anchor_generator",version="1"} 1729050.000000
nv_inference_request_duration_us{model="head_4",version="1"} 0.000000
nv_inference_request_duration_us{model="postprocess_detections",version="1"} 1024797.000000
nv_inference_request_duration_us{model="head_1",version="1"} 2007358.000000
nv_inference_request_duration_us{model="head_5",version="1"} 0.000000
nv_inference_request_duration_us{model="head_6",version="1"} 0.000000
nv_inference_request_duration_us{model="head_2",version="1"} 0.000000
nv_inference_request_duration_us{model="head_7",version="1"} 0.000000
nv_inference_request_duration_us{model="backbone",version="1"} 2524156.000000
nv_inference_request_duration_us{model="head_8",version="1"} 0.000000
nv_inference_request_duration_us{model="head_9",version="1"} 0.000000
nv_inference_request_duration_us{model="model_all_heads",version="1"} 0.000000
# HELP nv_inference_queue_duration_us Cumulative inference queuing duration in microseconds (includes cached requests)
# TYPE nv_inference_queue_duration_us counter
nv_inference_queue_duration_us{model="model_head_1",version="1"} 69.000000
nv_inference_queue_duration_us{model="head_3",version="1"} 0.000000
nv_inference_queue_duration_us{model="anchor_generator",version="1"} 8885.000000
nv_inference_queue_duration_us{model="head_4",version="1"} 0.000000
nv_inference_queue_duration_us{model="postprocess_detections",version="1"} 8371.000000
nv_inference_queue_duration_us{model="head_1",version="1"} 7996.000000
nv_inference_queue_duration_us{model="head_5",version="1"} 0.000000
nv_inference_queue_duration_us{model="head_6",version="1"} 0.000000
nv_inference_queue_duration_us{model="head_2",version="1"} 0.000000
nv_inference_queue_duration_us{model="head_7",version="1"} 0.000000
nv_inference_queue_duration_us{model="backbone",version="1"} 8433.000000
nv_inference_queue_duration_us{model="head_8",version="1"} 0.000000
nv_inference_queue_duration_us{model="head_9",version="1"} 0.000000
nv_inference_queue_duration_us{model="model_all_heads",version="1"} 0.000000
# HELP nv_inference_compute_input_duration_us Cumulative compute input duration in microseconds (does not include cached requests)
# TYPE nv_inference_compute_input_duration_us counter
nv_inference_compute_input_duration_us{model="model_head_1",version="1"} 2856970.000000
nv_inference_compute_input_duration_us{model="head_3",version="1"} 0.000000
nv_inference_compute_input_duration_us{model="anchor_generator",version="1"} 1328254.000000
nv_inference_compute_input_duration_us{model="head_4",version="1"} 0.000000
nv_inference_compute_input_duration_us{model="postprocess_detections",version="1"} 575610.000000
nv_inference_compute_input_duration_us{model="head_1",version="1"} 35120.000000
nv_inference_compute_input_duration_us{model="head_5",version="1"} 0.000000
nv_inference_compute_input_duration_us{model="head_6",version="1"} 0.000000
nv_inference_compute_input_duration_us{model="head_2",version="1"} 0.000000
nv_inference_compute_input_duration_us{model="head_7",version="1"} 0.000000
nv_inference_compute_input_duration_us{model="backbone",version="1"} 917843.000000
nv_inference_compute_input_duration_us{model="head_8",version="1"} 0.000000
nv_inference_compute_input_duration_us{model="head_9",version="1"} 0.000000
nv_inference_compute_input_duration_us{model="model_all_heads",version="1"} 0.000000
# HELP nv_inference_compute_infer_duration_us Cumulative compute inference duration in microseconds (does not include cached requests)
# TYPE nv_inference_compute_infer_duration_us counter
nv_inference_compute_infer_duration_us{model="model_head_1",version="1"} 4314026.000000
nv_inference_compute_infer_duration_us{model="head_3",version="1"} 0.000000
nv_inference_compute_infer_duration_us{model="anchor_generator",version="1"} 351166.000000
nv_inference_compute_infer_duration_us{model="head_4",version="1"} 0.000000
nv_inference_compute_infer_duration_us{model="postprocess_detections",version="1"} 428012.000000
nv_inference_compute_infer_duration_us{model="head_1",version="1"} 1951941.000000
nv_inference_compute_infer_duration_us{model="head_5",version="1"} 0.000000
nv_inference_compute_infer_duration_us{model="head_6",version="1"} 0.000000
nv_inference_compute_infer_duration_us{model="head_2",version="1"} 0.000000
nv_inference_compute_infer_duration_us{model="head_7",version="1"} 0.000000
nv_inference_compute_infer_duration_us{model="backbone",version="1"} 1582762.000000
nv_inference_compute_infer_duration_us{model="head_8",version="1"} 0.000000
nv_inference_compute_infer_duration_us{model="head_9",version="1"} 0.000000
nv_inference_compute_infer_duration_us{model="model_all_heads",version="1"} 0.000000
# HELP nv_inference_compute_output_duration_us Cumulative inference compute output duration in microseconds (does not include cached requests)
# TYPE nv_inference_compute_output_duration_us counter
nv_inference_compute_output_duration_us{model="model_head_1",version="1"} 49238.000000
nv_inference_compute_output_duration_us{model="head_3",version="1"} 0.000000
nv_inference_compute_output_duration_us{model="anchor_generator",version="1"} 38270.000000
nv_inference_compute_output_duration_us{model="head_4",version="1"} 0.000000
nv_inference_compute_output_duration_us{model="postprocess_detections",version="1"} 3656.000000
nv_inference_compute_output_duration_us{model="head_1",version="1"} 3375.000000
nv_inference_compute_output_duration_us{model="head_5",version="1"} 0.000000
nv_inference_compute_output_duration_us{model="head_6",version="1"} 0.000000
nv_inference_compute_output_duration_us{model="head_2",version="1"} 0.000000
nv_inference_compute_output_duration_us{model="head_7",version="1"} 0.000000
nv_inference_compute_output_duration_us{model="backbone",version="1"} 3781.000000
nv_inference_compute_output_duration_us{model="head_8",version="1"} 0.000000
nv_inference_compute_output_duration_us{model="head_9",version="1"} 0.000000
nv_inference_compute_output_duration_us{model="model_all_heads",version="1"} 0.000000
# HELP nv_cache_num_entries Number of responses stored in response cache
# TYPE nv_cache_num_entries gauge
# HELP nv_cache_num_lookups Number of cache lookups in response cache
# TYPE nv_cache_num_lookups gauge
# HELP nv_cache_num_hits Number of cache hits in response cache
# TYPE nv_cache_num_hits gauge
# HELP nv_cache_num_misses Number of cache misses in response cache
# TYPE nv_cache_num_misses gauge
# HELP nv_cache_num_evictions Number of cache evictions in response cache
# TYPE nv_cache_num_evictions gauge
# HELP nv_cache_lookup_duration Total cache lookup duration (hit and miss), in microseconds
# TYPE nv_cache_lookup_duration gauge
# HELP nv_cache_insertion_duration Total cache insertion duration, in microseconds
# TYPE nv_cache_insertion_duration gauge
# HELP nv_cache_util Cache utilization [0.0 - 1.0]
# TYPE nv_cache_util gauge
# HELP nv_cache_num_hits_per_model Number of cache hits per model
# TYPE nv_cache_num_hits_per_model counter
nv_cache_num_hits_per_model{model="model_head_1",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_3",version="1"} 0.000000
nv_cache_num_hits_per_model{model="anchor_generator",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_4",version="1"} 0.000000
nv_cache_num_hits_per_model{model="postprocess_detections",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_1",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_5",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_6",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_2",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_7",version="1"} 0.000000
nv_cache_num_hits_per_model{model="backbone",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_8",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_9",version="1"} 0.000000
nv_cache_num_hits_per_model{model="model_all_heads",version="1"} 0.000000
# HELP nv_cache_hit_lookup_duration_per_model Total cache hit lookup duration per model, in microseconds
# TYPE nv_cache_hit_lookup_duration_per_model counter
nv_cache_hit_lookup_duration_per_model{model="model_head_1",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_3",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="anchor_generator",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_4",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="postprocess_detections",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_1",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_5",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_6",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_2",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_7",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="backbone",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_8",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_9",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="model_all_heads",version="1"} 0.000000
# HELP nv_cache_num_misses_per_model Number of cache misses per model
# TYPE nv_cache_num_misses_per_model counter
nv_cache_num_misses_per_model{model="model_head_1",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_3",version="1"} 0.000000
nv_cache_num_misses_per_model{model="anchor_generator",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_4",version="1"} 0.000000
nv_cache_num_misses_per_model{model="postprocess_detections",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_1",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_5",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_6",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_2",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_7",version="1"} 0.000000
nv_cache_num_misses_per_model{model="backbone",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_8",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_9",version="1"} 0.000000
nv_cache_num_misses_per_model{model="model_all_heads",version="1"} 0.000000
# HELP nv_cache_miss_lookup_duration_per_model Total cache miss lookup duration per model, in microseconds
# TYPE nv_cache_miss_lookup_duration_per_model counter
nv_cache_miss_lookup_duration_per_model{model="model_head_1",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_3",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="anchor_generator",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_4",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="postprocess_detections",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_1",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_5",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_6",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_2",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_7",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="backbone",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_8",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_9",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="model_all_heads",version="1"} 0.000000
# HELP nv_cache_miss_insertion_duration_per_model Total cache miss insertion duration per model, in microseconds
# TYPE nv_cache_miss_insertion_duration_per_model counter
nv_cache_miss_insertion_duration_per_model{model="model_head_1",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_3",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="anchor_generator",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_4",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="postprocess_detections",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_1",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_5",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_6",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_2",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_7",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="backbone",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_8",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_9",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="model_all_heads",version="1"} 0.000000
# HELP nv_gpu_utilization GPU utilization rate [0.0 - 1.0)
# TYPE nv_gpu_utilization gauge
nv_gpu_utilization{gpu_uuid="GPU-2d050232-24ef-5168-4fc7-6700ac827bce"} 0.240000
# HELP nv_gpu_memory_total_bytes GPU total memory, in bytes
# TYPE nv_gpu_memory_total_bytes gauge
nv_gpu_memory_total_bytes{gpu_uuid="GPU-2d050232-24ef-5168-4fc7-6700ac827bce"} 25769803776.000000
# HELP nv_gpu_memory_used_bytes GPU used memory, in bytes
# TYPE nv_gpu_memory_used_bytes gauge
nv_gpu_memory_used_bytes{gpu_uuid="GPU-2d050232-24ef-5168-4fc7-6700ac827bce"} 16931356672.000000
# HELP nv_gpu_power_usage GPU power usage in watts
# TYPE nv_gpu_power_usage gauge
nv_gpu_power_usage{gpu_uuid="GPU-2d050232-24ef-5168-4fc7-6700ac827bce"} 156.564000
# HELP nv_gpu_power_limit GPU power management limit in watts
# TYPE nv_gpu_power_limit gauge
nv_gpu_power_limit{gpu_uuid="GPU-2d050232-24ef-5168-4fc7-6700ac827bce"} 350.000000
# HELP nv_energy_consumption GPU energy consumption in joules since the Triton Server started
# TYPE nv_energy_consumption counter
nv_energy_consumption{gpu_uuid="GPU-2d050232-24ef-5168-4fc7-6700ac827bce"} 3601.027000
```

In the second ensemble model with 9 heads, the prometheus metrics are like this :

```
# HELP nv_inference_request_success Number of successful inference requests, all batch sizes
# TYPE nv_inference_request_success counter
nv_inference_request_success{model="model_head_1",version="1"} 0.000000
nv_inference_request_success{model="head_3",version="1"} 100.000000
nv_inference_request_success{model="anchor_generator",version="1"} 100.000000
nv_inference_request_success{model="head_4",version="1"} 100.000000
nv_inference_request_success{model="postprocess_detections",version="1"} 900.000000
nv_inference_request_success{model="head_1",version="1"} 100.000000
nv_inference_request_success{model="head_5",version="1"} 100.000000
nv_inference_request_success{model="head_6",version="1"} 100.000000
nv_inference_request_success{model="head_2",version="1"} 100.000000
nv_inference_request_success{model="backbone",version="1"} 100.000000
nv_inference_request_success{model="head_7",version="1"} 100.000000
nv_inference_request_success{model="head_8",version="1"} 100.000000
nv_inference_request_success{model="head_9",version="1"} 100.000000
nv_inference_request_success{model="model_all_heads",version="1"} 100.000000
# HELP nv_inference_request_failure Number of failed inference requests, all batch sizes
# TYPE nv_inference_request_failure counter
nv_inference_request_failure{model="model_head_1",version="1"} 0.000000
nv_inference_request_failure{model="head_3",version="1"} 0.000000
nv_inference_request_failure{model="anchor_generator",version="1"} 0.000000
nv_inference_request_failure{model="head_4",version="1"} 0.000000
nv_inference_request_failure{model="postprocess_detections",version="1"} 0.000000
nv_inference_request_failure{model="head_1",version="1"} 0.000000
nv_inference_request_failure{model="head_5",version="1"} 0.000000
nv_inference_request_failure{model="head_6",version="1"} 0.000000
nv_inference_request_failure{model="head_2",version="1"} 0.000000
nv_inference_request_failure{model="backbone",version="1"} 0.000000
nv_inference_request_failure{model="head_7",version="1"} 0.000000
nv_inference_request_failure{model="head_8",version="1"} 0.000000
nv_inference_request_failure{model="head_9",version="1"} 0.000000
nv_inference_request_failure{model="model_all_heads",version="1"} 0.000000
# HELP nv_inference_count Number of inferences performed (does not include cached requests)
# TYPE nv_inference_count counter
nv_inference_count{model="model_head_1",version="1"} 0.000000
nv_inference_count{model="head_3",version="1"} 100.000000
nv_inference_count{model="anchor_generator",version="1"} 100.000000
nv_inference_count{model="head_4",version="1"} 100.000000
nv_inference_count{model="postprocess_detections",version="1"} 900.000000
nv_inference_count{model="head_1",version="1"} 100.000000
nv_inference_count{model="head_5",version="1"} 100.000000
nv_inference_count{model="head_6",version="1"} 100.000000
nv_inference_count{model="head_2",version="1"} 100.000000
nv_inference_count{model="backbone",version="1"} 100.000000
nv_inference_count{model="head_7",version="1"} 100.000000
nv_inference_count{model="head_8",version="1"} 100.000000
nv_inference_count{model="head_9",version="1"} 100.000000
nv_inference_count{model="model_all_heads",version="1"} 100.000000
# HELP nv_inference_exec_count Number of model executions performed (does not include cached requests)
# TYPE nv_inference_exec_count counter
nv_inference_exec_count{model="model_head_1",version="1"} 0.000000
nv_inference_exec_count{model="head_3",version="1"} 100.000000
nv_inference_exec_count{model="anchor_generator",version="1"} 100.000000
nv_inference_exec_count{model="head_4",version="1"} 100.000000
nv_inference_exec_count{model="postprocess_detections",version="1"} 900.000000
nv_inference_exec_count{model="head_1",version="1"} 100.000000
nv_inference_exec_count{model="head_5",version="1"} 100.000000
nv_inference_exec_count{model="head_6",version="1"} 100.000000
nv_inference_exec_count{model="head_2",version="1"} 100.000000
nv_inference_exec_count{model="backbone",version="1"} 100.000000
nv_inference_exec_count{model="head_7",version="1"} 100.000000
nv_inference_exec_count{model="head_8",version="1"} 100.000000
nv_inference_exec_count{model="head_9",version="1"} 100.000000
nv_inference_exec_count{model="model_all_heads",version="1"} 100.000000
# HELP nv_inference_request_duration_us Cumulative inference request duration in microseconds (includes cached requests)
# TYPE nv_inference_request_duration_us counter
nv_inference_request_duration_us{model="model_head_1",version="1"} 0.000000
nv_inference_request_duration_us{model="head_3",version="1"} 14311666.000000
nv_inference_request_duration_us{model="anchor_generator",version="1"} 5910979.000000
nv_inference_request_duration_us{model="head_4",version="1"} 14363945.000000
nv_inference_request_duration_us{model="postprocess_detections",version="1"} 58913533.000000
nv_inference_request_duration_us{model="head_1",version="1"} 14378334.000000
nv_inference_request_duration_us{model="head_5",version="1"} 14336454.000000
nv_inference_request_duration_us{model="head_6",version="1"} 14347145.000000
nv_inference_request_duration_us{model="head_2",version="1"} 14310618.000000
nv_inference_request_duration_us{model="backbone",version="1"} 2460564.000000
nv_inference_request_duration_us{model="head_7",version="1"} 14343129.000000
nv_inference_request_duration_us{model="head_8",version="1"} 14389983.000000
nv_inference_request_duration_us{model="head_9",version="1"} 14491176.000000
nv_inference_request_duration_us{model="model_all_heads",version="1"} 29087514.000000
# HELP nv_inference_queue_duration_us Cumulative inference queuing duration in microseconds (includes cached requests)
# TYPE nv_inference_queue_duration_us counter
nv_inference_queue_duration_us{model="model_head_1",version="1"} 0.000000
nv_inference_queue_duration_us{model="head_3",version="1"} 12510.000000
nv_inference_queue_duration_us{model="anchor_generator",version="1"} 10022.000000
nv_inference_queue_duration_us{model="head_4",version="1"} 11754.000000
nv_inference_queue_duration_us{model="postprocess_detections",version="1"} 46213640.000000
nv_inference_queue_duration_us{model="head_1",version="1"} 9433.000000
nv_inference_queue_duration_us{model="head_5",version="1"} 12068.000000
nv_inference_queue_duration_us{model="head_6",version="1"} 12850.000000
nv_inference_queue_duration_us{model="head_2",version="1"} 11972.000000
nv_inference_queue_duration_us{model="backbone",version="1"} 8074.000000
nv_inference_queue_duration_us{model="head_7",version="1"} 12434.000000
nv_inference_queue_duration_us{model="head_8",version="1"} 10090.000000
nv_inference_queue_duration_us{model="head_9",version="1"} 13365.000000
nv_inference_queue_duration_us{model="model_all_heads",version="1"} 116.000000
# HELP nv_inference_compute_input_duration_us Cumulative compute input duration in microseconds (does not include cached requests)
# TYPE nv_inference_compute_input_duration_us counter
nv_inference_compute_input_duration_us{model="model_head_1",version="1"} 0.000000
nv_inference_compute_input_duration_us{model="head_3",version="1"} 482239.000000
nv_inference_compute_input_duration_us{model="anchor_generator",version="1"} 4182645.000000
nv_inference_compute_input_duration_us{model="head_4",version="1"} 381857.000000
nv_inference_compute_input_duration_us{model="postprocess_detections",version="1"} 5768877.000000
nv_inference_compute_input_duration_us{model="head_1",version="1"} 384371.000000
nv_inference_compute_input_duration_us{model="head_5",version="1"} 443877.000000
nv_inference_compute_input_duration_us{model="head_6",version="1"} 616120.000000
nv_inference_compute_input_duration_us{model="head_2",version="1"} 438818.000000
nv_inference_compute_input_duration_us{model="backbone",version="1"} 918382.000000
nv_inference_compute_input_duration_us{model="head_7",version="1"} 576845.000000
nv_inference_compute_input_duration_us{model="head_8",version="1"} 444612.000000
nv_inference_compute_input_duration_us{model="head_9",version="1"} 708303.000000
nv_inference_compute_input_duration_us{model="model_all_heads",version="1"} 15347899.000000
# HELP nv_inference_compute_infer_duration_us Cumulative compute inference duration in microseconds (does not include cached requests)
# TYPE nv_inference_compute_infer_duration_us counter
nv_inference_compute_infer_duration_us{model="model_head_1",version="1"} 0.000000
nv_inference_compute_infer_duration_us{model="head_3",version="1"} 13609076.000000
nv_inference_compute_infer_duration_us{model="anchor_generator",version="1"} 1700379.000000
nv_inference_compute_infer_duration_us{model="head_4",version="1"} 13734642.000000
nv_inference_compute_infer_duration_us{model="postprocess_detections",version="1"} 6860768.000000
nv_inference_compute_infer_duration_us{model="head_1",version="1"} 13737362.000000
nv_inference_compute_infer_duration_us{model="head_5",version="1"} 13656193.000000
nv_inference_compute_infer_duration_us{model="head_6",version="1"} 13470334.000000
nv_inference_compute_infer_duration_us{model="head_2",version="1"} 13639434.000000
nv_inference_compute_infer_duration_us{model="backbone",version="1"} 1499234.000000
nv_inference_compute_infer_duration_us{model="head_7",version="1"} 13477953.000000
nv_inference_compute_infer_duration_us{model="head_8",version="1"} 13639134.000000
nv_inference_compute_infer_duration_us{model="head_9",version="1"} 13397781.000000
nv_inference_compute_infer_duration_us{model="model_all_heads",version="1"} 132423230.000000
# HELP nv_inference_compute_output_duration_us Cumulative inference compute output duration in microseconds (does not include cached requests)
# TYPE nv_inference_compute_output_duration_us counter
nv_inference_compute_output_duration_us{model="model_head_1",version="1"} 0.000000
nv_inference_compute_output_duration_us{model="head_3",version="1"} 197679.000000
nv_inference_compute_output_duration_us{model="anchor_generator",version="1"} 14340.000000
nv_inference_compute_output_duration_us{model="head_4",version="1"} 225947.000000
nv_inference_compute_output_duration_us{model="postprocess_detections",version="1"} 26516.000000
nv_inference_compute_output_duration_us{model="head_1",version="1"} 238129.000000
nv_inference_compute_output_duration_us{model="head_5",version="1"} 214396.000000
nv_inference_compute_output_duration_us{model="head_6",version="1"} 238709.000000
nv_inference_compute_output_duration_us{model="head_2",version="1"} 211102.000000
nv_inference_compute_output_duration_us{model="backbone",version="1"} 3424.000000
nv_inference_compute_output_duration_us{model="head_7",version="1"} 265899.000000
nv_inference_compute_output_duration_us{model="head_8",version="1"} 286779.000000
nv_inference_compute_output_duration_us{model="head_9",version="1"} 362117.000000
nv_inference_compute_output_duration_us{model="model_all_heads",version="1"} 2285966.000000
# HELP nv_cache_num_entries Number of responses stored in response cache
# TYPE nv_cache_num_entries gauge
# HELP nv_cache_num_lookups Number of cache lookups in response cache
# TYPE nv_cache_num_lookups gauge
# HELP nv_cache_num_hits Number of cache hits in response cache
# TYPE nv_cache_num_hits gauge
# HELP nv_cache_num_misses Number of cache misses in response cache
# TYPE nv_cache_num_misses gauge
# HELP nv_cache_num_evictions Number of cache evictions in response cache
# TYPE nv_cache_num_evictions gauge
# HELP nv_cache_lookup_duration Total cache lookup duration (hit and miss), in microseconds
# TYPE nv_cache_lookup_duration gauge
# HELP nv_cache_insertion_duration Total cache insertion duration, in microseconds
# TYPE nv_cache_insertion_duration gauge
# HELP nv_cache_util Cache utilization [0.0 - 1.0]
# TYPE nv_cache_util gauge
# HELP nv_cache_num_hits_per_model Number of cache hits per model
# TYPE nv_cache_num_hits_per_model counter
nv_cache_num_hits_per_model{model="model_head_1",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_3",version="1"} 0.000000
nv_cache_num_hits_per_model{model="anchor_generator",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_4",version="1"} 0.000000
nv_cache_num_hits_per_model{model="postprocess_detections",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_1",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_5",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_6",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_2",version="1"} 0.000000
nv_cache_num_hits_per_model{model="backbone",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_7",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_8",version="1"} 0.000000
nv_cache_num_hits_per_model{model="head_9",version="1"} 0.000000
nv_cache_num_hits_per_model{model="model_all_heads",version="1"} 0.000000
# HELP nv_cache_hit_lookup_duration_per_model Total cache hit lookup duration per model, in microseconds
# TYPE nv_cache_hit_lookup_duration_per_model counter
nv_cache_hit_lookup_duration_per_model{model="model_head_1",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_3",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="anchor_generator",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_4",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="postprocess_detections",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_1",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_5",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_6",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_2",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="backbone",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_7",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_8",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="head_9",version="1"} 0.000000
nv_cache_hit_lookup_duration_per_model{model="model_all_heads",version="1"} 0.000000
# HELP nv_cache_num_misses_per_model Number of cache misses per model
# TYPE nv_cache_num_misses_per_model counter
nv_cache_num_misses_per_model{model="model_head_1",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_3",version="1"} 0.000000
nv_cache_num_misses_per_model{model="anchor_generator",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_4",version="1"} 0.000000
nv_cache_num_misses_per_model{model="postprocess_detections",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_1",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_5",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_6",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_2",version="1"} 0.000000
nv_cache_num_misses_per_model{model="backbone",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_7",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_8",version="1"} 0.000000
nv_cache_num_misses_per_model{model="head_9",version="1"} 0.000000
nv_cache_num_misses_per_model{model="model_all_heads",version="1"} 0.000000
# HELP nv_cache_miss_lookup_duration_per_model Total cache miss lookup duration per model, in microseconds
# TYPE nv_cache_miss_lookup_duration_per_model counter
nv_cache_miss_lookup_duration_per_model{model="model_head_1",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_3",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="anchor_generator",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_4",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="postprocess_detections",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_1",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_5",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_6",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_2",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="backbone",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_7",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_8",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="head_9",version="1"} 0.000000
nv_cache_miss_lookup_duration_per_model{model="model_all_heads",version="1"} 0.000000
# HELP nv_cache_miss_insertion_duration_per_model Total cache miss insertion duration per model, in microseconds
# TYPE nv_cache_miss_insertion_duration_per_model counter
nv_cache_miss_insertion_duration_per_model{model="model_head_1",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_3",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="anchor_generator",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_4",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="postprocess_detections",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_1",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_5",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_6",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_2",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="backbone",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_7",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_8",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="head_9",version="1"} 0.000000
nv_cache_miss_insertion_duration_per_model{model="model_all_heads",version="1"} 0.000000
# HELP nv_gpu_utilization GPU utilization rate [0.0 - 1.0)
# TYPE nv_gpu_utilization gauge
nv_gpu_utilization{gpu_uuid="GPU-2d050232-24ef-5168-4fc7-6700ac827bce"} 0.000000
# HELP nv_gpu_memory_total_bytes GPU total memory, in bytes
# TYPE nv_gpu_memory_total_bytes gauge
nv_gpu_memory_total_bytes{gpu_uuid="GPU-2d050232-24ef-5168-4fc7-6700ac827bce"} 25769803776.000000
# HELP nv_gpu_memory_used_bytes GPU used memory, in bytes
# TYPE nv_gpu_memory_used_bytes gauge
nv_gpu_memory_used_bytes{gpu_uuid="GPU-2d050232-24ef-5168-4fc7-6700ac827bce"} 17082351616.000000
# HELP nv_gpu_power_usage GPU power usage in watts
# TYPE nv_gpu_power_usage gauge
nv_gpu_power_usage{gpu_uuid="GPU-2d050232-24ef-5168-4fc7-6700ac827bce"} 142.772000
# HELP nv_gpu_power_limit GPU power management limit in watts
# TYPE nv_gpu_power_limit gauge
nv_gpu_power_limit{gpu_uuid="GPU-2d050232-24ef-5168-4fc7-6700ac827bce"} 350.000000
# HELP nv_energy_consumption GPU energy consumption in joules since the Triton Server started
# TYPE nv_energy_consumption counter
nv_energy_consumption{gpu_uuid="GPU-2d050232-24ef-5168-4fc7-6700ac827bce"} 15561.756000
```

The problem is that there is several incoherence in the logs.
For example, in the first case the cumulative time on 100 requests of the inference request duration for the head_1 is
```
nv_inference_request_duration_us{model="head_1",version="1"} 2007358.000000
```
While in the second case of 9 heads, this time is
```
nv_inference_request_duration_us{model="head_1",version="1"} 14378334.000000
```
which is more than 7 times bigger (almost 9).
And I think it's just an error while computing metrics of ensemble model.

Another more concrete example.
On the second case with 9 heads, if we look at overall inference request duration and compute infer duration :
```
nv_inference_request_duration_us{model="model_all_heads",version="1"} 29087514.000000
nv_inference_compute_infer_duration_us{model="model_all_heads",version="1"} 132423230.000000
```
we see that the compute infer duration is far greater than the overall inference request duration.
Furthermore the mean client-side infererence time is around 700ms which is two times lower than the compute inference duration.

Thus I think the computation of the prometheus timing for ensemble models does not reflect the real timings.

**Triton Information**
I'm using the nvcr.io/nvidia/tritonserver:22.12-py3 docker image.

**To Reproduce**
An ensemble model where some tensorRT models (the heads) have the same inputs (from the backbone) and deliver their outputs to the same ONNX model (postprocess_detections).

**Expected behavior**
The metrics should be coherent.

**Comment by dyastremsky** [comment]
Thank you for your kind words, Julien. I appreciate you submitting such a detailed report. We have filed a ticket to investigate this issue.

We will likely need more from you in terms of full reproduction steps (even if on a smaller toy ensemble model or set-up), so if you can provide that, that'd be very helpful.

**Comment by julienripoche** [comment]
Hi !

Here is a toy model to reproduce the bug : 
[toy_model_0.zip](https://github.com/triton-inference-server/server/files/10609049/toy_model_0.zip)
[toy_model_1.zip](https://github.com/triton-inference-server/server/files/10609050/toy_model_1.zip)
[toy_model_2.zip](https://github.com/triton-inference-server/server/files/10609053/toy_model_2.zip)
[toy_model_3.zip](https://github.com/triton-inference-server/server/files/10609054/toy_model_3.zip)
[toy_model_4.zip](https://github.com/triton-inference-server/server/files/10609055/toy_model_4.zip)
Please concatenate these 5 files before unzipping.

I provide three ONNX models : 
- `backbone.onnx`
- `head.onnx`
- `concat.onnx`

And a model repository.

The two first model can be converted to ONNX with
```
trtexec --onnx=backbone.onnx --saveEngine=backbone.trt --inputIOFormats=fp16:chw --outputIOFormats=fp16:chw --fp16 --optShapes=image:1x3x2048x2048 --minShapes=image:1x3x256x256 --maxShapes=image:1x3x3072x2304
trtexec --onnx=head.onnx --saveEngine=head.trt --inputIOFormats=fp16:chw --outputIOFormats=fp16:chw --fp16 --optShapes=feature_1:1x256x256x256,feature_2:1x256x128x128,feature_3:1x256x64x64,feature_4:1x256x32x32,feature_5:1x256x16x16, --minShapes=feature_1:1x256x32x32,feature_2:1x256x16x16,feature_3:1x256x8x8,feature_4:1x256x4x4,feature_5:1x256x2x2 --maxShapes=feature_1:1x256x512x256,feature_2:1x256x256x128,feature_3:1x256x128x64,feature_4:1x256x64x32,feature_5:1x256x32x16
```

The ensemble model `ensemble_1` corresponds to the first case of my previous post and `ensemble_8` play the role of the second case.

The prometheus metrics are similar to the ones of the previous post.

**Comment by rmccorm4** [comment]
Hi @julienripoche,

Thanks for going into such depth. I'd like to take a deeper look into this. Would it be possible to share the most minimal setup required to reproduce? This would help me to investigate faster. 

For example:
- Maybe an ensemble composed of only 1 or 2 steps/models is required to reproduce
- Maybe only sending a single request rather than 100 can reproduce and would more easily show where the miscalculation is happening

**Comment by julienripoche** [comment]
Hi @rmccorm4,

Here is a simpler setup that reproduce at least parts of the wrong metrics.

Please export this pytorch model to ONNX:
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 6, 5)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 197 * 197, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)
    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = torch.flatten(x, 1) # flatten all dimensions except batch
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x
net = Net()
torch.onnx.export(
    net,
    torch.randn(1, 3, 800, 800),
    "simple_net.onnx",
    verbose=False,
    opset_version=11,
    input_names=["images"],
    output_names=["outputs"],
    dynamic_axes={"images": {0: "b"}, "outputs": {0: "b"}},
)
```

Place the ONNX model as `simple_net` model in this model repository:
[toy_model.zip](https://github.com/triton-inference-server/server/files/10968658/toy_model.zip)

Running one request on `ensemble_8` you should obtain a prometheus report like this:
```
# HELP nv_inference_request_success Number of successful inference requests, all batch sizes
# TYPE nv_inference_request_success counter
nv_inference_request_success{model="ensemble_1",version="1"} 0
nv_inference_request_success{model="ensemble_2",version="1"} 0
nv_inference_request_success{model="ensemble_8",version="1"} 1
nv_inference_request_success{model="ensemble_4",version="1"} 0
nv_inference_request_success{model="simple_net",version="1"} 8
# HELP nv_inference_request_failure Number of failed inference requests, all batch sizes
# TYPE nv_inference_request_failure counter
nv_inference_request_failure{model="ensemble_1",version="1"} 0
nv_inference_request_failure{model="ensemble_2",version="1"} 0
nv_inference_request_failure{model="ensemble_8",version="1"} 0
nv_inference_request_failure{model="ensemble_4",version="1"} 0
nv_inference_request_failure{model="simple_net",version="1"} 0
# HELP nv_inference_count Number of inferences performed (does not include cached requests)
# TYPE nv_inference_count counter
nv_inference_count{model="ensemble_1",version="1"} 0
nv_inference_count{model="ensemble_2",version="1"} 0
nv_inference_count{model="ensemble_8",version="1"} 1
nv_inference_count{model="ensemble_4",version="1"} 0
nv_inference_count{model="simple_net",version="1"} 8
# HELP nv_inference_exec_count Number of model executions performed (does not include cached requests)
# TYPE nv_inference_exec_count counter
nv_inference_exec_count{model="ensemble_1",version="1"} 0
nv_inference_exec_count{model="ensemble_2",version="1"} 0
nv_inference_exec_count{model="ensemble_8",version="1"} 1
nv_inference_exec_count{model="ensemble_4",version="1"} 0
nv_inference_exec_count{model="simple_net",version="1"} 8
# HELP nv_inference_request_duration_us Cumulative inference request duration in microseconds (includes cached requests)
# TYPE nv_inference_request_duration_us counter
nv_inference_request_duration_us{model="ensemble_1",version="1"} 0
nv_inference_request_duration_us{model="ensemble_2",version="1"} 0
nv_inference_request_duration_us{model="ensemble_8",version="1"} 337748
nv_inference_request_duration_us{model="ensemble_4",version="1"} 0
nv_inference_request_duration_us{model="simple_net",version="1"} 2618832
# HELP nv_inference_queue_duration_us Cumulative inference queuing duration in microseconds (includes cached requests)
# TYPE nv_inference_queue_duration_us counter
nv_inference_queue_duration_us{model="ensemble_1",version="1"} 0
nv_inference_queue_duration_us{model="ensemble_2",version="1"} 0
nv_inference_queue_duration_us{model="ensemble_8",version="1"} 2
nv_inference_queue_duration_us{model="ensemble_4",version="1"} 0
nv_inference_queue_duration_us{model="simple_net",version="1"} 2281292
# HELP nv_inference_compute_input_duration_us Cumulative compute input duration in microseconds (does not include cached requests)
# TYPE nv_inference_compute_input_duration_us counter
nv_inference_compute_input_duration_us{model="ensemble_1",version="1"} 0
nv_inference_compute_input_duration_us{model="ensemble_2",version="1"} 0
nv_inference_compute_input_duration_us{model="ensemble_8",version="1"} 27825
nv_inference_compute_input_duration_us{model="ensemble_4",version="1"} 0
nv_inference_compute_input_duration_us{model="simple_net",version="1"} 27821
# HELP nv_inference_compute_infer_duration_us Cumulative compute inference duration in microseconds (does not include cached requests)
# TYPE nv_inference_compute_infer_duration_us counter
nv_inference_compute_infer_duration_us{model="ensemble_1",version="1"} 0
nv_inference_compute_infer_duration_us{model="ensemble_2",version="1"} 0
nv_inference_compute_infer_duration_us{model="ensemble_8",version="1"} 309498
nv_inference_compute_infer_duration_us{model="ensemble_4",version="1"} 0
nv_inference_compute_infer_duration_us{model="simple_net",version="1"} 309494
# HELP nv_inference_compute_output_duration_us Cumulative inference compute output duration in microseconds (does not include cached requests)
# TYPE nv_inference_compute_output_duration_us counter
nv_inference_compute_output_duration_us{model="ensemble_1",version="1"} 0
nv_inference_compute_output_duration_us{model="ensemble_2",version="1"} 0
nv_inference_compute_output_duration_us{model="ensemble_8",version="1"} 95
nv_inference_compute_output_duration_us{model="ensemble_4",version="1"} 0
nv_inference_compute_output_duration_us{model="simple_net",version="1"} 91
# HELP nv_cache_num_hits_per_model Number of cache hits per model
# TYPE nv_cache_num_hits_per_model counter
nv_cache_num_hits_per_model{model="ensemble_1",version="1"} 0
nv_cache_num_hits_per_model{model="ensemble_2",version="1"} 0
nv_cache_num_hits_per_model{model="ensemble_8",version="1"} 0
nv_cache_num_hits_per_model{model="ensemble_4",version="1"} 0
nv_cache_num_hits_per_model{model="simple_net",version="1"} 0
# HELP nv_cache_hit_lookup_duration_per_model Total cache hit lookup duration per model, in microseconds
# TYPE nv_cache_hit_lookup_duration_per_model counter
nv_cache_hit_lookup_duration_per_model{model="ensemble_1",version="1"} 0
nv_cache_hit_lookup_duration_per_model{model="ensemble_2",version="1"} 0
nv_cache_hit_lookup_duration_per_model{model="ensemble_8",version="1"} 0
nv_cache_hit_lookup_duration_per_model{model="ensemble_4",version="1"} 0
nv_cache_hit_lookup_duration_per_model{model="simple_net",version="1"} 0
# HELP nv_cache_num_misses_per_model Number of cache misses per model
# TYPE nv_cache_num_misses_per_model counter
nv_cache_num_misses_per_model{model="ensemble_1",version="1"} 0
nv_cache_num_misses_per_model{model="ensemble_2",version="1"} 0
nv_cache_num_misses_per_model{model="ensemble_8",version="1"} 0
nv_cache_num_misses_per_model{model="ensemble_4",version="1"} 0
nv_cache_num_misses_per_model{model="simple_net",version="1"} 0
# HELP nv_cache_miss_lookup_duration_per_model Total cache miss lookup duration per model, in microseconds
# TYPE nv_cache_miss_lookup_duration_per_model counter
nv_cache_miss_lookup_duration_per_model{model="ensemble_1",version="1"} 0
nv_cache_miss_lookup_duration_per_model{model="ensemble_2",version="1"} 0
nv_cache_miss_lookup_duration_per_model{model="ensemble_8",version="1"} 0
nv_cache_miss_lookup_duration_per_model{model="ensemble_4",version="1"} 0
nv_cache_miss_lookup_duration_per_model{model="simple_net",version="1"} 0
# HELP nv_cache_miss_insertion_duration_per_model Total cache miss insertion duration per model, in microseconds
# TYPE nv_cache_miss_insertion_duration_per_model counter
nv_cache_miss_insertion_duration_per_model{model="ensemble_1",version="1"} 0
nv_cache_miss_insertion_duration_per_model{model="ensemble_2",version="1"} 0
nv_cache_miss_insertion_duration_per_model{model="ensemble_8",version="1"} 0
nv_cache_miss_insertion_duration_per_model{model="ensemble_4",version="1"} 0
nv_cache_miss_insertion_duration_per_model{model="simple_net",version="1"} 0
# HELP nv_energy_consumption GPU energy consumption in joules since the Triton Server started
# TYPE nv_energy_consumption counter
nv_energy_consumption{gpu_uuid="GPU-e923826c-9faa-c23f-f7d3-fce2ad1ad8c4"} 897.505
# HELP nv_gpu_utilization GPU utilization rate [0.0 - 1.0)
# TYPE nv_gpu_utilization gauge
nv_gpu_utilization{gpu_uuid="GPU-e923826c-9faa-c23f-f7d3-fce2ad1ad8c4"} 0
# HELP nv_gpu_memory_total_bytes GPU total memory, in bytes
# TYPE nv_gpu_memory_total_bytes gauge
nv_gpu_memory_total_bytes{gpu_uuid="GPU-e923826c-9faa-c23f-f7d3-fce2ad1ad8c4"} 11811160064
# HELP nv_gpu_memory_used_bytes GPU used memory, in bytes
# TYPE nv_gpu_memory_used_bytes gauge
nv_gpu_memory_used_bytes{gpu_uuid="GPU-e923826c-9faa-c23f-f7d3-fce2ad1ad8c4"} 1605369856
# HELP nv_gpu_power_usage GPU power usage in watts
# TYPE nv_gpu_power_usage gauge
nv_gpu_power_usage{gpu_uuid="GPU-e923826c-9faa-c23f-f7d3-fce2ad1ad8c4"} 50.032
# HELP nv_gpu_power_limit GPU power management limit in watts
# TYPE nv_gpu_power_limit gauge
nv_gpu_power_limit{gpu_uuid="GPU-e923826c-9faa-c23f-f7d3-fce2ad1ad8c4"} 260
# HELP nv_cpu_utilization CPU utilization rate [0.0 - 1.0]
# TYPE nv_cpu_utilization gauge
nv_cpu_utilization 0.01880484747179273
# HELP nv_cpu_memory_total_bytes CPU total memory (RAM), in bytes
# TYPE nv_cpu_memory_total_bytes gauge
nv_cpu_memory_total_bytes 67128946688
# HELP nv_cpu_memory_used_bytes CPU used memory (RAM), in bytes
# TYPE nv_cpu_memory_used_bytes gauge
nv_cpu_memory_used_bytes 21045710848
```

Where these times
```
nv_inference_request_duration_us{model="ensemble_8",version="1"} 337748
nv_inference_request_duration_us{model="simple_net",version="1"} 2618832
```
are incoherents as ensemble_8 is only composed of simple_net.

Furthermore, these times
```
nv_inference_request_duration_us{model="simple_net",version="1"} 2618832
nv_inference_queue_duration_us{model="simple_net",version="1"} 2281292
```
are larger than the client side request time of around 330ms.

**Comment by rmccorm4** [comment]
Hi @julienripoche,

Thanks for sharing the simplified example with only a single request. After going over the numbers and your ensemble config, I think these numbers make sense.

You are sending 1 request to your ensemble_8 model, which in turn produces 8 requests sent concurrently to your simple_net model.

This can be seen in the inference counts:
```
# 1 inference done on top level model
nv_inference_request_success{model="ensemble_8",version="1"} 1
# 8 inferences done on composing model
nv_inference_request_success{model="simple_net",version="1"} 8
```

To back that up, we can see the inference time spent on the composing model is about 8x the reported ensemble inference (because each inference is done in parallel in this example):
```
nv_inference_request_duration_us{model="ensemble_8",version="1"} 337748
nv_inference_request_duration_us{model="simple_net",version="1"} 2618832
```
Working with some rough numbers here:
- `2618832 us / 8 inferences == 327354 us/inf`, so `327354us == average infer time on a single simple_net request`
- `327354us + some overhead to schedule composing ensemble models and gather ensemble outputs into response ~= 337748us`

It looks like the majority of this time is actually spent with requests in `queue` as seen by :
```
nv_inference_queue_duration_us{model="ensemble_8",version="1"} 2
nv_inference_queue_duration_us{model="simple_net",version="1"} 2281292
```

So the non-queue time remaining for the simple_net model is `2618832 - 2281292 == 337540`, which looks like it roughly adds up to the sum of the `compute_*` times for simple_net, as expected. Since these requests have been generated and scheduled by the ensemble model already, but are sitting in the queue for `simple_net`, they are considered in queue time for `simple_net` but not for the ensemble. This is why we see an almost zero queue time for the ensemble model.

For the `compute_*` times, the ensemble model itself doesn't actually do much computation at all, it mostly just schedules the requests to its composing models and gathers the inputs/outputs, so the numbers are expected to be very close to the composing models numbers here.

---

So to summarize, I don't think the numbers you are seeing here are wrong. The numbers you want to be comparing to your actual perceived e2e request latency / client latency would be the numbers reported by the ensemble model. The composing model of the ensemble (simple_net), since in this example is just the same model used repeatedly, will report much higher sum totals since the model is called repeatedly internally.

**Comment by rmccorm4** [comment]
Hi @julienripoche, I'm going to go ahead and close this for now. If you have further questions, please follow up or open a new issue.
